src/BSON/Value.c

```diff
@@ -0,0 +1,631 @@
+}
+
+static PHP_METHOD(MongoDB_BSON_Value, getDecimal128)
+{
+	php_phongo_value_t* intern;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern = Z_VALUE_OBJ_P(getThis());
+
+	PHONGO_VALUE_CHECK_TYPE(intern->value, BSON_TYPE_DECIMAL128);
+
+	php_phongo_bson_new_decimal128(return_value, &intern->value.value.v_decimal128);
+}
+
+static PHP_METHOD(MongoDB_BSON_Value, getDocument)
+{
+	php_phongo_value_t* intern;
+	bson_t*             bson;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern = Z_VALUE_OBJ_P(getThis());
+
+	PHONGO_VALUE_CHECK_TYPE(intern->value, BSON_TYPE_DOCUMENT);
+
+	bson = bson_new_from_data(intern->value.value.v_doc.data, intern->value.value.v_doc.data_len);
+
+	php_phongo_bson_new_document(return_value, bson, false);
+}
+
+static PHP_METHOD(MongoDB_BSON_Value, getDouble)
+{
+	php_phongo_value_t* intern;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern = Z_VALUE_OBJ_P(getThis());
+
+	PHONGO_VALUE_CHECK_TYPE(intern->value, BSON_TYPE_DOUBLE);
+
+	RETURN_DOUBLE(intern->value.value.v_double);
+}
+
+static PHP_METHOD(MongoDB_BSON_Value, getInt)
+{
+	php_phongo_value_t* intern;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern = Z_VALUE_OBJ_P(getThis());
+
+	switch (intern->value.value_type) {
+		case BSON_TYPE_INT32:
+			RETURN_LONG(intern->value.value.v_int32);
+
+		case BSON_TYPE_INT64:
+			RETURN_INT64(intern->value.value.v_int64);
+
+		default:
+			phongo_throw_exception(PHONGO_ERROR_UNEXPECTED_VALUE, "Unexpected BSON type %d", intern->value.value_type);
+	}
+}
+
+static PHP_METHOD(MongoDB_BSON_Value, getInt32)
+{
+	php_phongo_value_t* intern;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern = Z_VALUE_OBJ_P(getThis());
+
+	PHONGO_VALUE_CHECK_TYPE(intern->value, BSON_TYPE_INT32);
+
+	RETURN_LONG(intern->value.value.v_int32);
+}
+
+static PHP_METHOD(MongoDB_BSON_Value, getInt64)
+{
+	php_phongo_value_t* intern;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern = Z_VALUE_OBJ_P(getThis());
+
+	PHONGO_VALUE_CHECK_TYPE(intern->value, BSON_TYPE_INT64);
+
+	RETURN_INT64(intern->value.value.v_int64);
+}
+
+static PHP_METHOD(MongoDB_BSON_Value, getMaxKey)
+{
+	php_phongo_value_t* intern;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern = Z_VALUE_OBJ_P(getThis());
+
+	PHONGO_VALUE_CHECK_TYPE(intern->value, BSON_TYPE_MAXKEY);
+
+	object_init_ex(return_value, php_phongo_maxkey_ce);
+}
+
+static PHP_METHOD(MongoDB_BSON_Value, getMinKey)
+{
+	php_phongo_value_t* intern;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern = Z_VALUE_OBJ_P(getThis());
+
+	PHONGO_VALUE_CHECK_TYPE(intern->value, BSON_TYPE_MINKEY);
+
+	object_init_ex(return_value, php_phongo_minkey_ce);
+}
+
+static PHP_METHOD(MongoDB_BSON_Value, getNull)
+{
+	php_phongo_value_t* intern;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern = Z_VALUE_OBJ_P(getThis());
+
+	PHONGO_VALUE_CHECK_TYPE(intern->value, BSON_TYPE_NULL);
+
+	RETURN_NULL();
+}
+
+static PHP_METHOD(MongoDB_BSON_Value, getNumber)
```

Comment:
I'm not sure if `getNumber` makes a lot of sense in PHP, but I figured I'd add it anyways for convenience.